### DISCUSSION TITLE
Roadmap | F.A.Q.

### DISCUSSION THREAD

**Discussion by ggerganov** [ggml-org/whisper.cpp#126] [Q&A]
## Updated roadmap as Github Project

https://github.com/users/ggerganov/projects/7

## Roadmap (old)

In decreasing priority:

- [x] [Decoding strategies](https://github.com/ggerganov/whisper.cpp/pull/291)
  Try to achieve at least parity with the OpenAI implementation
  Target release: [v1.1.0](https://github.com/ggerganov/whisper.cpp/releases/tag/v1.1.0)

- [x] [Memory usage reduction](https://github.com/ggerganov/whisper.cpp/pull/431)
  This will allow wider application on low-memory devices. Should be possible to cut memory usage in half with a few simple changes in `ggml`
  Target release: [v1.2.0](https://github.com/ggerganov/whisper.cpp/releases/tag/v1.2.0)

- [x] [Core ML support](https://github.com/ggerganov/whisper.cpp/pull/566)
  This will allow to utilize the Apple Neural Engine for very efficient inference of the model
  Target release: [v1.3.0](https://github.com/ggerganov/whisper.cpp/releases/tag/v1.3.0)

- [x] [Q4 / Q5 / Q8 integer quantization](https://github.com/ggerganov/whisper.cpp/discussions/281)
  Added via https://github.com/ggerganov/whisper.cpp/pull/540
  Target release: v1.4.0

- [x] GPU support
  [Partial CUDA support via cuBLAS](https://github.com/ggerganov/whisper.cpp/pull/834)
  Target release: v1.4.0

- [ ] [Documentation of ggml](https://github.com/ggerganov/whisper.cpp/issues/40)
  Hopefully this leads to more contributions

- [ ] [Diarization](https://github.com/ggerganov/whisper.cpp/issues/64)
  Very wanted feature, but very difficult to achieve. Interesting to [explore and experiment](https://github.com/ggerganov/whisper.cpp/pull/130)

- [ ] [Low-power mode](https://github.com/ggerganov/whisper.cpp/issues/200#issuecomment-1337564222)
  This has the potential for some performance improvements too

## F.A.Q.

- ### Is `whisper.cpp` faster or slower than `PyTorch CPU`?

  The performance should be comparable. At the time of writing, the performance on Apple Silicon when using `whisper.cpp` is better since we utilise FP16 + Accelerate framework, while `PyTorch` [not yet](https://github.com/openai/whisper/discussions/51). But this will soon change.

  In general, it is not very easy to make a proper benchmark between the 2 implementations. For more information, read the following comment: https://github.com/ggerganov/whisper.cpp/issues/89#issuecomment-1352233809

- ### Should I use `whisper.cpp` in my project?

  `whisper.cpp` is a hobby project. It does not strive to provide a production ready implementation. The main goals of the implementation is to be educational, minimalistic, portable, hackable and performant. There are no guarantees that the implementation is correct and bug-free and stuff can break at any point in the future. Support and updates will depend mostly on contributions, since with time I will move on and won't dedicate too much time on the project.

  If you plan to use `whisper.cpp` in your own project, keep in mind the above.
  My advice is to not put all your eggs into the `whisper.cpp` basket.

- ### How can I contribute?

  - [Add tests](https://github.com/ggerganov/whisper.cpp/blob/master/tests/run-tests.sh)
  - [Add bindings](https://github.com/ggerganov/whisper.cpp#bindings)
  - [Add examples](https://github.com/ggerganov/whisper.cpp/tree/master/examples)
  - [Add benchmarks](https://github.com/ggerganov/whisper.cpp/issues/89)
  - [Good first issues](https://github.com/ggerganov/whisper.cpp/issues?q=is%3Aopen+is%3Aissue+label%3A%22good+first+issue%22)
  - [Help wanted issues](https://github.com/ggerganov/whisper.cpp/issues?q=is%3Aopen+is%3Aissue+label%3A%22help+wanted%22)
  - [Experiment](https://github.com/ggerganov/whisper.cpp/issues/137)
  - [Make cool demos](https://github.com/ggerganov/whisper.cpp/discussions/167)

- ### Will `ggml` / `whisper.cpp` support CUDA / GPU?

  One of the main goals of this implementation is to be very minimalistic and be able to run it on a large spectrum of hardware. The existing CPU-only implementation achieves this goal - it is bloat-free and very simple. I think it also has some educational value. Of course, not taking advantage of modern GPU hardware is a huge drawback in terms of performance. However, adding a dependency on a certain GPU framework will tie the project with the corresponding hardware and will introduce some extra complexity.
  
  With that said, adding GPU support to the project is low priority.

  In any case, it would not be too difficult to add initial support. The main thing that needs to be offloaded to the GPU is the `GGML_OP_MUL_MAT` operator:
  https://github.com/ggerganov/whisper.cpp/blob/c71363f14cb0d31efe5c74f148b268da649050d9/ggml.c#L6231-L6234
  This is where more than 90% of the computation time is currently spent. Also, I don't think it's necessary to offload the entire model to the GPU. For example, the 2 convolution layers at the start of the Encoder can easily remain on the CPU as they are not very computationally heavy. Not uploading the full model to VRAM will make it require less memory and thus make it compatible with more video cards.

  Another candidate GPU framework that will likely be supported in the future is Apple's [Metal Performance Shaders (MPS)](https://developer.apple.com/documentation/metalperformanceshaders?language=objc). Currently, `ggml` supports Apple's [Accelerate](https://developer.apple.com/documentation/accelerate) framework and I really like how seamlessly it integrates in the project - both on MacOS and iOS. It does not feel like a third-party dependency at all and that is why I think MPS support can be added in a similar way. In theory, it will help to utilize the GPU on Apple devices and could potentially lead to some additional performance improvement. Also, the unified memory model of modern Apple Silicon devices allows to seamlessly share the model weights and the data embeddings between the CPU and the GPU, which is not the case for CUDA. So far, [my initial experiments haven't shown any benefit of using MPS for the transformer inference](https://github.com/ggerganov/ggml/tree/master/examples/gpt-j#attempt-to-use-the-m1-gpu), but maybe some more work is needed.

  Edit: Sample Metal support has been demonstrated here: https://github.com/ggerganov/whisper.cpp/pull/127

  Edit2: There is a promising CUDA support in the works through NVBLAS: https://github.com/ggerganov/whisper.cpp/issues/220

  Edit3: CUDA support via cuBLAS has been added: https://github.com/ggerganov/whisper.cpp/pull/834

## Notable `whisper.cpp` discussions

  - [Original announcement](https://github.com/openai/whisper/discussions/130)
  - [Reddit /r/MachineLearning](https://www.reddit.com/r/MachineLearning/comments/y0nvqu/p_pure_cc_port_of_openais_whisper/)
  - [Hackaday](https://hackaday.com/2022/11/27/heres-a-plain-c-c-implementation-of-ai-speech-recognition-so-get-hackin/)
  - [Hacker News (Show HN)](https://news.ycombinator.com/item?id=33877893)
  - [The New Yorker](https://www.newyorker.com/tech/annals-of-technology/whispers-of-ais-modular-future)
  - [Changelog podcast](https://changelog.com/podcast/532)

**Comment by rpikus27** [comment]
I saw the demo with command recognition. How do you trigger events after a command? Is there an intent handler model built in?

**Reply by ggerganov** [reply]
The `command` example only demonstrates how to receive voice commands. Handling the commands is up to you.
You can either extend the C++ example with a handler or you can pipe / parse the output from an external app, etc.
There is no built-in intent model available.

I am planning to soon demonstrate how to extend the `command` example for the purpose of  making moves on a chess board with voice. Updates will be posted in #428 if you are interested.

**Reply by rpikus27** [reply]
I am interested what you choose to parse the text. Whisper is so good, for my purposes, a switch statement would do the trick, but a mass market app will probably need a model.

**Reply by ruban22** [reply]
Hi,

Im intersted in your solution for real time streaming transcription/translation in our Video confrencing,  can you offer your service to implement it in our enviroment?


Thanks

**Comment by katurov** [comment]
Hi! I've built version for M1 as in instruction but after all I have my stream translated from original language to English. And I cannot switch it off. Is here some ideas how to?

**Reply by SonOfLilit** [reply]
Found the solution here:

https://github.com/ggerganov/whisper.cpp/discussions/526

```
./main --language auto ...
```

**Comment by fromparis** [comment]
can you make an example of the word time stamp in swiftUI ?? I'm totally lost in c++

**Comment by rexendevar** [comment]
Is there any current or planned method of setting up a file to use as an additional dictionary?

**Comment by gamavin** [comment]
Hello,

I'm experiencing a persistent JNI crash in my Android app when trying to load ggml-small.bin and ggml-small-q8_0.bin models using a recent version of whisper.cpp and ggml (latest from GitHub master).

Key Issues:

Crash Point: The app crashes consistently inside whisper_init_from_file_with_params, specifically at ggml_backend_dev_backend_reg (offset +12 in the function, or +288 from Java_com_undefault_alma_data_remote_WhisperBridge_transcribe).

system_info Output: The whisper_print_system_info() in Android Logcat only shows WHISPER : COREML = 0 | OPENVINO = 0 |. It does not show NEON = 1, even though I'm targeting aarch64-none-linux-android29 and have tried enabling NEON in CMakeLists.txt with:

set(CMAKE_C_FLAGS "${CMAKE_C_FLAGS} -march=armv8-a+simd")
set(CMAKE_CXX_FLAGS "${CMAKE_CXX_FLAGS} -march=armv8-a+simd")
target_compile_definitions(whisper_jni PRIVATE GGML_NEON=1)

Build: The native library (libwhisper_jni.so) is confirmed to be recompiling (BuildId changes after clean rebuilds).

Environment:

Android NDK: 27.0.12077973

Target ABI: aarch64-none-linux-android29 (tested on Galaxy S24 Ultra)

Model used: ggml-small.bin (non-quantized) and ggml-small-q8_0.bin

Logcat Snippet (Crash & System Info):

I/WhisperBridge: 🔍 v1.1 Memuat model dari: /data/user/0/com.package.name/files/ggml-small.bin
I/WhisperBridge: ℹ️ Informasi Sistem Whisper:
                  WHISPER : COREML = 0 | OPENVINO = 0 | 
I/WhisperBridge: 🧠 Inisialisasi context dengan backend CPU...
A/DEBUG: #00 pc 0000000000xxxxxx  /data/app/.../libwhisper_jni.so (offset 0xyyyyyyyy) (ggml_backend_dev_backend_reg+12) (BuildId: new_build_id)
A/DEBUG: #06 pc 0000000000zzzzzz  /data/app/.../libwhisper_jni.so (offset 0xyyyyyyyy) (Java_com_package_name_WhisperBridge_transcribe+288) (BuildId: new_build_id)

(Please replace com.package.name, xxxxxx, yyyyyyyy, zzzzzz, and new_build_id with your actual values from Logcat)

Is this a known issue with recent ggml versions when using .bin models on Android, or is there a specific configuration/compilation flag I might be missing to correctly initialize the CPU backend and enable NEON detection/reporting?

Any guidance would be greatly appreciated. Thank you!

**Comment by tech-Anuranjith** [comment]
how can i make a translation instead of transcription in android app using this example repo.

thanks in advance
